drgn/helpers/linux/slab.py

```diff
@@ -244,3 +252,70 @@ def _slab_page_objects(page: Object, slab: Object) -> Iterator[Object]:
         slab = cast(slab_type, page)
         if slab.slab_cache == slab_cache:
             yield from _slab_page_objects(page, slab)
+
+
+@overload
+def find_containing_slab_cache(addr: Object) -> Object:
+    """
+    Get the slab cache that an address (given as an :class:`.Object`) was
+    allocated from, if any.
+    :param addr: Address to look up.
+    :return: ``struct kmem_cache *`` containing *addr*, or ``NULL`` if *addr*
+        is not from a slab cache.
+    """
+    ...
+
+
+@overload
+def find_containing_slab_cache(prog: Program, addr: IntegerLike) -> Object:
+    """
+    Get the slab cache that an address (given as a :class:`.Program` and an
+    integer) was allocated from, if any.
+
+    Note that SLOB does not store enough information to identify objects in a
+    slab cache, so if the kernel is configured to use SLOB, this will always
+    return ``NULL``.
+
+    :param addr: Address to look up.
+    :return: ``struct kmem_cache *`` containing *addr*, or ``NULL`` if *addr*
+        is not from a slab cache.
+    """
+    ...
+
+
+def find_containing_slab_cache(  # type: ignore  # Need positional-only arguments.
+    prog_or_addr: Union[Program, Object], addr: Optional[IntegerLike] = None
+) -> Object:
+    if addr is None:
+        assert isinstance(prog_or_addr, Object)
+        prog = prog_or_addr.prog_
+        addr = prog_or_addr
+    else:
+        assert isinstance(prog_or_addr, Program)
+        prog = prog_or_addr
+    addr = operator.index(addr)
+
+    start_addr = pfn_to_virt(prog["min_low_pfn"]).value_()
+    end_addr = (pfn_to_virt(prog["max_pfn"]) + prog["PAGE_SIZE"]).value_()
+
+    if addr < start_addr or addr >= end_addr:
+        # Not a directly mapped address
+        return NULL(prog, "struct kmem_cache *")
+
+    page = virt_to_page(prog, addr)
+
+    try:
+        PG_slab_mask = 1 << prog.constant("PG_slab")
+        if not page.flags & PG_slab_mask:
+            # Not a slab page
+            return NULL(prog, "struct kmem_cache *")
+
+    except FaultError:
+        # Page does not exist
+        return NULL(prog, "struct kmem_cache *")
```

Comment:
We can make the try/except even more specific:

```suggestion
    try:
        page_flags = page.flags.read_()
    except FaultError:
        # Page does not exist
        return NULL(prog, "struct kmem_cache *")
    if not (page_flags & (1 << prog.constant("PG_slab"))):
        # Not a slab page
        return NULL(prog, "struct kmem_cache *")
```

Reply:
Moved!